added support for arbitrary measurements to snapshots

JuanGiraldo0212 opened the pull request:
**Context:**

Add support for arbitrary measurements to Snapshot

**Description of the Change:**

PennyLane currently has the qml.Snapshot operation that works in conjunction with the qml.snapshots transform to allow users to inspect the state at any point in a calculation. But, the state can be hard to interpret and analyze for larger numbers of qubits, limiting its utility in sanity checks and debugging.

To fix this, we want to support arbitrary state-based measurements with snapshots. 

**Benefits:**

Allows users to debug and analyse their quantum circuits in a more intuitive and undestandable manner.

**Possible Drawbacks:**

**Related GitHub Issues:**

#4876

albi3ro commented on tests/test_debugging.py:
```diff
@@ -332,3 +332,48 @@ def qnode(params):
+        result = qml.snapshots(circuit)(0.1)
+        expected = {
+            0: np.array(1),
+            1: np.array(0.00996671),
+            2: np.array([0.99003329, 0.00996671]),
+            3: np.array([0.98877108 + 0.0j, 0.0 - 0.14943813j]),
+            "execution_results": np.array(0.92106099),
```
An optional suggestion.

Can we compare against analytic expectations (ie `np.sin(x/2)**2` or whatever) instead of hardcoded numbers?

Tests serve to increase our confidence that the output is correct. If someone else comes back to this test at a later point trying to debug something that's gone wrong, how hard is it for them to determine whether the current results are wrong or the hard-coded values were wrong?

JuanGiraldo0212 replied:
That makes, sense. I will implement the proposed changes. Thanks!

albi3ro reviewed (commented):
This PR looks great 🎉 

Don't forget to add your name your name in the "Contributors" list near the bottom of the changelog.

Two non-blocking comments:
1) As mentioned below, comparing against analytic results instead of hard coded numbers
2) Adding some unit tests to:

https://github.com/PennyLaneAI/pennylane/blob/d7fe50a3305a686cd750b3e286bb1c31b50142cb/tests/devices/qubit/test_apply_operation.py#L538C20-L538C20

Sorry I forgot about that testing spot till now. This will allow us to test just the `apply_operation` component in isolation.

trbromley commented on doc/releases/changelog-dev.md:
```diff
@@ -149,6 +149,9 @@
   Additionally, the relevant tests have been organized and placed in a new file, `tests/ops/test_meta.py` .
   [(#4789)](https://github.com/PennyLaneAI/pennylane/pull/4789)
 
+* The function ``qml.Snapshot`` now supports arbitrary measurements of type ``StateMeasurement``.
+  [(#4876)](https://github.com/PennyLaneAI/pennylane/pull/4908)
+
```
Feel free to add yourself to the list of contributors at the bottom :smile:

lillian542 commented:
Hi @JuanGiraldo0212 - just a quick suggestion. I notice you are force-pushing your changes as you update the PR. Would you mind using `git commit` instead? It's a bit cleaner and easier to read through the changes for the PR review that way. Thanks for working on this!

JuanGiraldo0212 commented:
> Hi @JuanGiraldo0212 - just a quick suggestion. I notice you are force-pushing your changes as you update the PR. Would you mind using `git commit` instead? It's a bit cleaner and easier to read through the changes for the PR review that way. Thanks for working on this!

Oh sorry I missed this comment, sure I will start doing that.

JuanGiraldo0212 commented:
> This PR looks great 🎉
> 
> Don't forget to add your name your name in the "Contributors" list near the bottom of the changelog.
> 
> Two non-blocking comments:
> 
> 1. As mentioned below, comparing against analytic results instead of hard coded numbers
> 2. Adding some unit tests to:
> 
> https://github.com/PennyLaneAI/pennylane/blob/d7fe50a3305a686cd750b3e286bb1c31b50142cb/tests/devices/qubit/test_apply_operation.py#L538C20-L538C20
> 
> Sorry I forgot about that testing spot till now. This will allow us to test just the `apply_operation` component in isolation.

I have added myself to the list of contributors and added the new test, however I'm still a little confused on how to implement the analytic expectations. Not sure if I have to come up with a function that returns the same value that I have hardcoded when I pass the argument x. If so, could you maybe provide a snippet were there is a similar approach. Thanks!

trbromley commented on pennylane/ops/meta.py:
```diff
@@ -156,7 +156,8 @@ class Snapshot(Operation):
     execution steps of a quantum function. As such, it is a pseudo operation
-    with no effect on the quantum state.
+    with no effect on the quantum state. Arbitrary measurements are supported
+    in snapshots via the key word argument ``measurement``.
```
```suggestion
    in snapshots via the keyword argument ``measurement``.
```

trbromley commented on pennylane/ops/meta.py:
```diff
@@ -167,6 +168,9 @@ class Snapshot(Operation):
         tag (str or None): An optional custom tag for the snapshot, used to index it
             in the snapshots dictionary.
 
+        measurement (StateMeasurement or None): An optional option to record arbitrary
+            measurements of a state.
```
```suggestion
        measurement (StateMeasurement or None): An optional argument to record arbitrary
            measurements of a state.
```

trbromley commented on pennylane/ops/meta.py:
```diff
@@ -167,6 +168,9 @@ class Snapshot(Operation):
         tag (str or None): An optional custom tag for the snapshot, used to index it
             in the snapshots dictionary.
 
+        measurement (StateMeasurement or None): An optional option to record arbitrary
```
Are we restricted to state-based measurement processes only?

albi3ro replied:
For now at least. We might be able to open it up to sample-based measurements in the future, but that would require a few more changes.

trbromley commented on pennylane/debugging.py:
Please could you also make the update here: https://xanaduai-pennylane--4908.com.readthedocs.build/en/4908/introduction/inspecting_circuits.html#debugging-with-mid-circuit-snapshots

albi3ro reviewed (approved):
🎉 Looking great :) This PR was very nicely done and I am happy to give it an approval. I hope you enjoyed the experience too :)

As for the random failures in the `test_sampling.py` case, we've been seeing that randomly happen on PR's over the last week.  We have yet to figure out what is going on with those tests and what has changed lately, but usually you can rerun the tests and it will go away 🤷

JuanGiraldo0212 commented:
> 🎉 Looking great :) This PR was very nicely done and I am happy to give it an approval. I hope you enjoyed the experience too :)
> 
> As for the random failures in the `test_sampling.py` case, we've been seeing that randomly happen on PR's over the last week. We have yet to figure out what is going on with those tests and what has changed lately, but usually you can rerun the tests and it will go away 🤷

Ohh I see, I thought that test failing was my fault haha. Thanks for approving the PR, and yes I really enjoyed working on Pennylane, it's always nice to work on very well structured projects. Just one question how can I re-run the tests? I don't see an option to do it on the GUI. Have a nice day @albi3ro.

albi3ro commented:
I just tried rerunning it by clicking "Details" next to the failing test, and then clicking the circle with arrows in the upper right corner.  

But that didn't seem to work this time.  I might recommend retriggering everything instead.  That can be done by doing something like:
```
git commit -m "trigger ci" --allow-empty
git push
```

locally with an empty commit.  

🤞

albi3ro commented:
@JuanGiraldo0212 I just merged something into master and updated your PR that might help.

JuanGiraldo0212 commented:
> @JuanGiraldo0212 I just merged something into master and updated your PR that might help.

Thanks! Now I see there is another error, I haven't seen that one before, not sure if it is also related to what you mentioned before or if I should fix something.

lillian542 reviewed (approved):
Thanks @JuanGiraldo0212 , looks great! That last test that is failing is a stochastic one that fails occasionally, I just restarted it so it should be passing soon! 🤞

Alex-Preciado commented:
Awesome!! Now that this PR has been merged, I'll go ahead and close https://github.com/PennyLaneAI/pennylane/issues/4876. Thank you so much @JuanGiraldo0212 😎
